API versioning using accept header

Marishka17 opened the pull request:
<!---
Copyright (C) 2020-2022 Intel Corporation

SPDX-License-Identifier: MIT
-->

<!-- Raised an issue to propose your change (https://github.com/opencv/cvat/issues).
It helps to avoid duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [CONTRIBUTION](https://github.com/opencv/cvat/blob/develop/CONTRIBUTING.md)
guide. -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
<!-- Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here. Describe your changes in detail, add
screenshots. -->
This PR contains a new API versioning scheme using [AcceptHeaderVersioning](https://www.django-rest-framework.org/api-guide/versioning/#acceptheaderversioning)
I've tried to implement an approach like [API versioning for HTTP REST interfaces (best practices)](https://gist.github.com/kirushik/cbe130d2bd4847020c6fe5e953e92618#use-versioned-vnd-content-types) (`application/vnd.cvat.v1.0+json`) but in this case, I couldn't fix unavailable drf-spectacular documentation. I guess It may be not supported in the current [implementation](https://github.dev/tfranzel/drf-spectacular/blob/803c5c12ee7e0ecfce79c4875ff2dd96e7e060bf/drf_spectacular/plumbing.py#L905-L906)
### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable by a reason then ~~explicitly strikethrough~~ the whole
line. If you don't do that github will show an incorrect process for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I submit my changes into the `develop` branch
- [ ] I have added a description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2022 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

nmanovic commented on .github/workflows/schedule.yml:
```diff
@@ -16,10 +16,10 @@ jobs:
         run: |
           docker-compose -f docker-compose.yml -f docker-compose.dev.yml -f ./tests/docker-compose.email.yml -f tests/docker-compose.file_share.yml -f components/serverless/docker-compose.serverless.yml up -d --build
-          /bin/bash -c 'while [[ "$(curl -s -o /dev/null -w ''%{http_code}'' ${API_ABOUT_PAGE})" != "401" ]]; do sleep 5; done'
+          /bin/bash -c 'while [[ $(curl -s -o /dev/null -H "Accept: application/vnd.cvat+json; version=1.0" -w "%{http_code}" ${API_ABOUT_PAGE}) != "401" ]]; do sleep 5; done'
```
@Marishka17 , what do you think if by default we will use the latest version of API? Thus if we don't specify accept header, DRF will use the latest one. I know that the link https://github.com/interagent/http-api-design/blob/master/en/foundations/require-versioning-in-the-accepts-header.md says that it is a bad practice. But GitHub uses the strategy: https://docs.github.com/en/rest/overview/media-types

Marishka17 replied:
@nmanovic ,  Also rest_framework [documentation](https://www.django-rest-framework.org/api-guide/versioning/) doesn't say that this is a bad practice, let's use the default version. In this case, I guess we can miss `accept header` in client/tests. I'll update it.

nmanovic commented on CHANGELOG.md:
```diff
@@ -35,6 +35,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Changed
 - Users don't have access to a task object anymore if they are assigneed only on some jobs of the task (<https://github.com/openvinotoolkit/cvat/pull/3788>)
 - Different resources (tasks, projects) are not visible anymore for all CVAT instance users by default (<https://github.com/openvinotoolkit/cvat/pull/3788>)
+- API versioning scheme: using accept header versioning instead of namespace versioning (<https://github.com/openvinotoolkit/cvat/pull/4239>)
```
did we use URL versioning? https://www.django-rest-framework.org/api-guide/versioning/#urlpathversioning

nmanovic replied:
OK. It is namespace versioning: <i>to the client, this scheme is the same as URLPathVersioning. The only difference is how it is configured in your Django application, as it uses URL namespacing, instead of URL keyword arguments.</i>

nmanovic commented on cvat/apps/dataset_manager/tests/test_formats.py:
```diff
@@ -36,24 +37,30 @@ def generate_image_file(filename, size=(100, 100)):
 
 class ForceLogin:
-    def __init__(self, user, client):
+    def __init__(self, user, client, version=settings.BACKEND_VERSIONS.V1_0):
```
@Marishka17 , One more problem if we don't have a default version. I don't think that it is good idea to have BACKEND_VERSIONS.V1_0 and ACCEPT_HEADER_TEMPLATE in settings.

nmanovic commented on cvat/apps/engine/versioning.py:
```diff
@@ -0,0 +1,26 @@
+    Accept: application/vnd.cvat+json; version=1.0
+    """
+
+    def determine_version(self, request, *args, **kwargs):
```
@Marishka17 , Do we need the method if we have a default version?

nmanovic commented on cvat/settings/base.py:
```diff
@@ -154,9 +169,10 @@ def add_ssh_keys():
+        'cvat.apps.engine.versioning.CustomAcceptHeaderVersioning',
     # Need to add 'api-docs' here as a workaround for include_docs_urls.
-    'ALLOWED_VERSIONS': ('v1', 'api-docs'),
+    'ALLOWED_VERSIONS': (*BACKEND_VERSIONS.list(), 'api-docs', 'downloading'),
```
Let's try to remove `api-docs` and `downloading` from the list.

nmanovic commented on site/content/en/docs/contributing/rest-api-design.md:
```diff
@@ -36,7 +36,7 @@ Common scheme for our REST API is `<VERB> [namespace] <objects> <id> <action>`.
 - Maintain good security practices
 - Cache data to improve performance
-- Versioning our APIs (e.g. `/api/v1`, `/api/v2`). It should be done when you
+- Versioning our APIs (e.g. `/api`, `/api/v2`). It should be done when you
```
@Marishka17 , need to update the statement. It looks wrong now (e.g., `/api/v2`)

nmanovic commented on tests/cypress/support/commands_organizations.js:
```diff
@@ -31,9 +33,10 @@ Cypress.Commands.add('createOrganization', (organizationParams) => {
+        url: '/api/organizations?page_size=all',
         headers: {
             Authorization: `Token ${authKey}`,
+            Accept: acceptHeader,
```
@Marishka17 , @bsekachev , should UI work with the latest version (=default) of REST API always? Thus we can skip acceptHeader for tests and UI.

nmanovic commented on tests/rest_api/test_0001_users_api.py:
```diff
@@ -12,12 +12,14 @@
 def test_non_admin_cannot_see_others():
     for username in ['dummy1', 'worker1', 'user1', 'business1']:
-        response = requests.get(config.get_api_url('users'), auth=(username, config.USER_PASS))
+        response = requests.get(config.get_api_url('users'), auth=(username, config.USER_PASS),
```
@Marishka17 , tests should test all versions. Thus if we have a default version, right now it will be enough. Let's discuss. Really want to find the right approach.

nmanovic commented on utils/cli/cli.py:
```diff
@@ -34,6 +34,8 @@ def main():
         api = CVAT_API_V1('%s:%s' % (args.server_host, args.server_port), args.https)
         cli = CLI(session, api, args.auth)
+        #TODO add version indication
+        session.headers.update({'Accept': 'application/vnd.cvat+json; version=1.0'})
```
Again, probably CLI should work with the latest (default version) of API.

Marishka17 commented:
@nmanovic , I've updated the API version to 2.0 as you said yesterday (I didn't do it before because I believed that we should have done tasks with server code redesigning and separating it to standalone applications).
Also, I've added a test for checking the version for `server/about` endpoint.
I've found several ways to version codebase (simply extend serializer, an approach like migrations, and others) and I guess we should discuss this moment with the team to choose the best approach for us. So It should be another task.

nmanovic commented on cvat/settings/base.py:
```diff
@@ -151,12 +155,10 @@ def add_ssh_keys():
-    # Need to add 'api-docs' here as a workaround for include_docs_urls.
-    'ALLOWED_VERSIONS': ('v1', 'api-docs'),
+        'rest_framework.versioning.AcceptHeaderVersioning',
+    'ALLOWED_VERSIONS': ('1.0', '2.0'),
```
@Marishka17 , do we support 1.0?

nmanovic reviewed (approved):
LGTM

Marishka17 commented:
@nmanovic, Should we also [register](https://www.iana.org/form/media-types) vnd MIME type?

nmanovic commented:
@Marishka17 , we should. Need to contact OSPDT.
